client/src/services/DonuService.ts

```diff
@@ -300,3 +300,124 @@ export const getModelResult = async (
     console.error('[DONU Service] — getModelResult', `No ${selectedModelGraphType} Graph available in this model`); // eslint-disable-line no-console
   }
 };
+
+const L2Norm = (x: number[]): number => {
```

Comment:
All those utility function should literally be in a _utils_ file out of this service.